examples/src/main/scala/basic/GPT2.scala

```diff
@@ -1,378 +1,405 @@
+      val tLength = t.shape(Axis[Heads |*| L])
+      require(tLength % numHeads == 0, s"T length $tLength not divisible by numHeads $numHeads")
+      t.rearrange(
+        (Axis[Heads], Axis[L]),
+        (Axis[Heads] -> numHeads, Axis[L] -> (tLength / numHeads))
+      )
+    val qkvLength = cAttn.weight.shape(Axis[QKV])
+    require(qkvLength % 3 == 0, s"QKV length $qkvLength not divisible by 3")
+    val (qLength, kLength, vLength) = (qkvLength / 3, qkvLength / 3, qkvLength / 3)
+
+    val wq = cAttn.weight.slice(Axis[QKV] -> (0 until qLength)).relabel(Axis[QKV] -> Axis[Heads |*| Query])
+    val wkb = cAttn.bias.slice(Axis[QKV] -> (qLength until qLength + kLength)).relabel(Axis[QKV] -> Axis[Heads |*| Key])
+    val wk = cAttn.weight.slice(Axis[QKV] -> (qLength until qLength + kLength)).relabel(Axis[QKV] -> Axis[Heads |*| Key])
+    val wqb = cAttn.bias.slice(Axis[QKV] -> (0 until qLength)).relabel(Axis[QKV] -> Axis[Heads |*| Query])
+    val wv = cAttn.weight.slice(Axis[QKV] -> (qLength + kLength until qkvLength)).relabel(Axis[QKV] -> Axis[Heads |*| HeadValue])
+    val wvb = cAttn.bias.slice(Axis[QKV] -> (qLength + kLength until qkvLength)).relabel(Axis[QKV] -> Axis[Heads |*| HeadValue])
+
+    MultiHeadAttentionParams(
+      wq = HeadsParams(splitWeightToHeads(wq, numHeads), splitBiasToHeads(wqb, numHeads)),
+      wk = HeadsParams(splitWeightToHeads(wk, numHeads), splitBiasToHeads(wkb, numHeads)),
+      wv = HeadsParams(splitWeightToHeads(wv, numHeads), splitBiasToHeads(wvb, numHeads)),
+      proj = cProj
+    )
+
+case class EmbeddingMixerParams(
+    c_fc: LinearLayerParams[Embedding, EmbeddingMixer],
+    c_proj: LinearLayerParams[EmbeddingMixer, Embedding]
+)
+
+type WTEParams = Tensor2[Vocab, Embedding, Float]
+type WPEParams = Tensor2[Context, Embedding, Float]
+
+case class TransformerLayerParams(
+    ln1: LayerNormalizationParams,
+    attn: MultiHeadAttentionParams,
+    ln2: LayerNormalizationParams,
+    embeddingMixer: EmbeddingMixerParams
+)
+
+case class GPT2Params(
+    vocabularyEmbeddings: WTEParams,
+    positionalEmbeddings: WPEParams,
+    layers: List[TransformerLayerParams],
+    outputNormalization: LayerNormalizationParams,
+    output: ProjectionLayerParams[Embedding, Vocab]
+)
+
+object GPT2Params:
+  def apply(
+      positionalEmbeddings: WPEParams,
+      vocabularyEmbeddings: WTEParams,
+      layers: List[TransformerLayerParams],
+      outputNormalization: LayerNormalizationParams
+  ): GPT2Params =
+    val outputParams = ProjectionLayerParams(
+      vocabularyEmbeddings.transpose // Tying output weights with input embeddings
+    )
+    GPT2Params(vocabularyEmbeddings, positionalEmbeddings, layers, outputNormalization, outputParams)
+
+case class GPT2(params: GPT2Params):
+
+  private case class LinearLayer[In: Label, Out: Label](params: LinearLayerParams[In, Out]) extends (Tensor1[In, Float] => Tensor1[Out, Float]):
```

Comment:
To me the extends Function does not add much information and makes everything look heavy.  All the type information is already in the apply signature.  Should we drop the extends?